framework/src/base/NonlinearSystem.C

```diff
@@ -1786,16 +1794,49 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
       {
         _fe_problem.reinitNodeFace(node, boundary_id, 0);
 
-        for (std::vector<NodalBC *>::const_iterator it = _bcs[0].getNodalBCs(boundary_id).begin(); it != _bcs[0].getNodalBCs(boundary_id).end(); ++it)
+        const std::vector<NodalBC *> & bcs = _bcs[0].getNodalBCs(boundary_id);
+        for (std::vector<NodalBC *>::const_iterator it = bcs.begin(); it != bcs.end(); ++it)
         {
           NodalBC * bc = *it;
-          if (bc->shouldApply() && bc->variable().isNodalDefined())
-            zero_rows.push_back(bc->variable().nodalDofIndex());
+
+          // Get the list of variables to which this BC is actually
+          // coupled.  We should not need to call
+          // computeOffDiagJacobian() for variables the BC is not
+          // coupled to.  Note: this list does not contain bc->variable.number().
+          const std::vector<MooseVariable *> & coupled_moose_vars = bc->getCoupledMooseVars();
+
+          // Extract set of coupled MooseVariable numbers, also
+          // sneakily insert bc->variable.number() to make the logic
+          // below a bit simpler.
+          std::set<unsigned int> involved_moose_var_nums;
+          for (unsigned int var=0; var < coupled_moose_vars.size(); ++var)
```

Comment:
I tried to figure out how you want to do this, but I couldn't.

Are you wanting to invert the order of the loops, which are currently:
```
for (each boundary node)
  for (each NodalBC)
```

to
 
```
for (each NodalBC)
  for (each boundary node)
```

?

Because whether or not a given variable is coupled to the BC depends on each BC, so the check needs to be inside the BC loop.  The only way the check can be done outside the node loop is if the node loop is inside the BC loop...